[SQL] RustError: `cast_to_dN_fN` expected `Option<F32>` but found `Option<F64>`

**There is an extra cast to double from real on the outside when the real has already been cast to double to multiply with the given decimal value.**

SQL:

```sql
CREATE TABLE t1(c0 REAL) with ('materialized' = 'true');

CREATE MATERIALIZED VIEW v3_optimized AS SELECT t1.c0 * 0.42331 FROM t1;
```

The SQL compilation succeeds but Rust fails with:

```rust
Rust error: the Rust code generated based on the SQL (possibly combined with any user-provided UDF Rust and TOML code) failed to compile.
This should not happen (except if the error is in user-provided UDF code).
Please file a bug report with the example SQL and any UDF Rust/TOML that triggers it at:
https://github.com/feldera/feldera/issues


The compilation task exited with status code 101 and produced the following logs:

stderr:
   Compiling project-0b7bb52ef43c2a55c548456cc63cd709ea801a8daf5b2d25c7c70b0702906446 v0.1.0 (/home/abhizer/.feldera/compiler/rust-compilation/projects/project-0b7bb52ef43c2a55c548456cc63cd709ea801a8daf5b2d25c7c70b0702906446)
error[E0308]: mismatched types
   --> projects/project-0b7bb52ef43c2a55c548456cc63cd709ea801a8daf5b2d25c7c70b0702906446/src/main.rs:127:37
    |
127 |             Tup1::new(cast_to_dN_fN(times_dN_d(cast_to_dN_fN((*t_1).0), cast_to_d_decimal(new_decimal("0.42331", 6, 5).unwrap()))))
    |                       ------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<F32>`, found `Option<F64>`
    |                       |
    |                       arguments to this function are incorrect
    |
    = note: expected enum `std::option::Option<feldera_sqllib::F32>`
               found enum `std::option::Option<feldera_sqllib::F64>`
note: function defined here
   --> /mnt/hdd0/Development/feldera.git/main/crates/sqllib/src/casts.rs:681:8
    |
681 | pub fn cast_to_dN_fN(value: Option<F32>) -> Option<F64> {
    |        ^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `project-0b7bb52ef43c2a55c548456cc63cd709ea801a8daf5b2d25c7c70b0702906446` (bin "project-0b7bb52ef43c2a55c548456cc63cd709ea801a8daf5b2d25c7c70b0702906446") due to 1 previous error


stdout:
```

--- a/crates/sqllib/src/lib.rs
+++ b/crates/sqllib/src/lib.rs
@@ -877,38 +877,53 @@ pub fn wrap_bool(b: Option<bool>) -> bool {
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn or_b_b(left: bool, right: bool) -> bool {
-    left || right
+pub fn or_b_b<F>(left: bool, right: F) -> bool
+where
+    F: Fn() -> bool,
+{
+    left || right()
 }
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn or_bN_b(left: Option<bool>, right: bool) -> Option<bool> {
-    match (left, right) {
-        (Some(l), r) => Some(l || r),
-        (_, true) => Some(true),
-        (_, _) => None::<bool>,
+pub fn or_bN_b<F>(left: Option<bool>, right: F) -> Option<bool>
+where
+    F: Fn() -> bool,
+{
+    match left {
+        Some(l) => Some(l || right()),
+        None => match right() {
+            true => Some(true),
+            _ => None,
+        },
     }
 }
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn or_b_bN(left: bool, right: Option<bool>) -> Option<bool> {
-    match (left, right) {
-        (l, Some(r)) => Some(l || r),
-        (true, _) => Some(true),
-        (_, _) => None::<bool>,
+pub fn or_b_bN<F>(left: bool, right: F) -> Option<bool>
+where
+    F: Fn() -> Option<bool>,
+{
+    match left {
+        false => right(),
+        true => Some(true),
     }
 }
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn or_bN_bN(left: Option<bool>, right: Option<bool>) -> Option<bool> {
-    match (left, right) {
-        (Some(l), Some(r)) => Some(l || r),
-        (Some(true), _) => Some(true),
-        (_, Some(true)) => Some(true),
-        (_, _) => None::<bool>,
+pub fn or_bN_bN<F>(left: Option<bool>, right: F) -> Option<bool>
+where
+    F: Fn() -> Option<bool>,
+{
+    match left {
+        None => match right() {
+            Some(true) => Some(true),
+            _ => None,
+        },
+        Some(false) => right(),
+        Some(true) => Some(true),
     }
 }
 
@@ -916,38 +931,54 @@ pub fn or_bN_bN(left: Option<bool>, right: Option<bool>) -> Option<bool> {
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn and_b_b(left: bool, right: bool) -> bool {
-    left && right
+pub fn and_b_b<F>(left: bool, right: F) -> bool
+where
+    F: Fn() -> bool,
+{
+    left && right()
 }
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn and_bN_b(left: Option<bool>, right: bool) -> Option<bool> {
-    match (left, right) {
-        (Some(l), r) => Some(l && r),
-        (_, false) => Some(false),
-        (_, _) => None::<bool>,
+pub fn and_bN_b<F>(left: Option<bool>, right: F) -> Option<bool>
+where
+    F: Fn() -> bool,
+{
+    match left {
+        Some(false) => Some(false),
+        Some(true) => Some(right()),
+        None => match right() {
+            false => Some(false),
+            _ => None,
+        },
     }
 }
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn and_b_bN(left: bool, right: Option<bool>) -> Option<bool> {
-    match (left, right) {
-        (l, Some(r)) => Some(l && r),
-        (false, _) => Some(false),
-        (_, _) => None::<bool>,
+pub fn and_b_bN<F>(left: bool, right: F) -> Option<bool>
+where
+    F: Fn() -> Option<bool>,
+{
+    match left {
+        false => Some(false),
+        true => right(),
     }
 }
 
 #[doc(hidden)]
 #[inline(always)]
-pub fn and_bN_bN(left: Option<bool>, right: Option<bool>) -> Option<bool> {
-    match (left, right) {
-        (Some(l), Some(r)) => Some(l && r),
-        (Some(false), _) => Some(false),
-        (_, Some(false)) => Some(false),
-        (_, _) => None::<bool>,
+pub fn and_bN_bN<F>(left: Option<bool>, right: F) -> Option<bool>
+where
+    F: Fn() -> Option<bool>,
+{
+    match left {
+        Some(false) => Some(false),
+        Some(true) => right(),
+        None => match right() {
+            Some(false) => Some(false),
+            _ => None,
+        },
     }
 }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
@@ -930,6 +930,22 @@ public class ToRustInnerVisitor extends InnerVisitor {
                 this.builder.append(")");
                 break;
             }
+            case OR:
+            case AND: {
+                String function = RustSqlRuntimeLibrary.INSTANCE.getFunctionName(
+                        expression.getNode(),
+                        expression.opcode,
+                        expression.getType(),
+                        expression.left.getType(),
+                        expression.right.getType());
+                this.builder.append(function).append("(");
+                expression.left.accept(this);
+                // lazy in the right argument, make it a closure
+                this.builder.append(", || (");
+                expression.right.accept(this);
+                this.builder.append("))");
+                break;
+            }
             default: {
                 String function = RustSqlRuntimeLibrary.INSTANCE.getFunctionName(
                         expression.getNode(),

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
@@ -362,10 +362,10 @@ public class ToRustVisitor extends CircuitVisitor {
     }
 
     void generateOperator(DBSPOperator operator) {
-        /*
-        String str = operator.getNode().toInternalString();
-        this.writeComments(str);
-         */
+        if (this.compiler.options.ioOptions.verbosity > 0) {
+            String str = operator.getNode().toInternalString();
+            this.writeComments(str);
+        }
         operator.accept(this);
         this.builder.newline();
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -324,8 +324,8 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                     return left.withMayBeNull(anyNull);
             }
             if (rd != null) {
-                // FLOAT op DECIMAL, convert to FLOAT
-                return lf;
+                // FLOAT op DECIMAL, convert to DOUBLE
+                return new DBSPTypeDouble(left.getNode(), anyNull);
             }
         }
         if (ld != null) {
@@ -335,8 +335,8 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                         Math.max(ln.getPrecision(), rn.getPrecision()), ld.scale, anyNull);
             }
             if (rf != null) {
-                // DECIMAL op FLOAT, convert to FLOAT
-                return rf;
+                // DECIMAL op FLOAT, convert to DOUBLE
+                return new DBSPTypeDouble(right.getNode(), anyNull);
             }
             // DECIMAL op DECIMAL does not convert to a common type.
         }
@@ -441,6 +441,7 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                     opcode == DBSPOpcode.SUB || opcode == DBSPOpcode.MUL) {
                 // Use the inferred Calcite type for the output as the common type
                 commonBase = typeWithNull;
+                expressionResultType = commonBase;
             }
             if (opcode.isComparison()) {
                 expressionResultType = DBSPTypeBool.create(anyNull);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/MinMaxOptimize.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/MinMaxOptimize.java
@@ -124,13 +124,14 @@ public class MinMaxOptimize extends Passes {
 
             DBSPVariablePath inputVar = new DBSPTypeTuple(aggregationInputType).ref().var();
             DBSPClosureExpression init = new DBSPTupleExpression(
-                    inputVar.deref().field(0).cast(resultType))
+                    inputVar.deref().field(0).applyCloneIfNeeded().cast(resultType))
                     .closure(inputVar, this.weightVar);
 
             DBSPVariablePath acc = new DBSPTypeTuple(resultType).var();
             DBSPClosureExpression comparison =
                     new DBSPTupleExpression(new DBSPBinaryExpression(operator.getNode(),
-                            resultType, code, acc.field(0), inputVar.deref().field(0))
+                            resultType, code, acc.field(0).applyCloneIfNeeded(),
+                            inputVar.deref().field(0).applyCloneIfNeeded())
                             .cast(resultType))
                             .closure(acc, inputVar, this.weightVar);
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -1159,4 +1159,19 @@ public class RegressionTests extends SqlIoTest {
         visitor.apply(circuit);
         this.compileRustTestCase(sql);
     }
+
+    @Test
+    public void issue3093() {
+        this.compileRustTestCase("""
+                CREATE TABLE t1(c0 REAL);
+                CREATE VIEW v0 AS
+                SELECT t1.c0 + 0.42331, t1.c0 * 0.42331, .42331 * t1.c0 FROM t1;""");
+    }
+
+    @Test
+    public void issue3094() {
+        this.compileRustTestCase("""
+                CREATE TABLE t1(c0 REAL, c1 VARCHAR, c2 CHAR);
+                CREATE VIEW v2_optimized AS SELECT MIN(t1.c1) FROM t1 WHERE (1 IS NOT DISTINCT FROM 2);""");
+    }
 }

--- a/sql-to-dbsp-compiler/slt/src/main/java/org/dbsp/sqllogictest/executors/DBSPExecutor.java
+++ b/sql-to-dbsp-compiler/slt/src/main/java/org/dbsp/sqllogictest/executors/DBSPExecutor.java
@@ -596,6 +596,7 @@ public class DBSPExecutor extends SqlSltTestExecutor {
                 compilerOptions.languageOptions.throwOnError = options.stopAtFirstError;
                 compilerOptions.languageOptions.lenient = true;
                 compilerOptions.languageOptions.generateInputForEveryTable = true;
+                compilerOptions.ioOptions.verbosity = options.verbosity;
                 DBSPExecutor result = new DBSPExecutor(options, compilerOptions);
                 result.skip(skip.get());
                 Set<String> bugs = options.readBugsFile();

